Simbody/src/GeneralForceSubsystem.cpp

```diff
@@ -34,12 +34,230 @@
+                    } else { // ordinary velocity dependent force
+                        impl.calcForce(*m_state, *m_rigidBodyForces, *m_particleForces, *m_mobilityForces);
+                    }
+                }
+            } else {
+                // Process a single parallel force. Subtract 1 from threadIndex b/c
+                // we use 0 for the non-parallel forces.
+                const auto& impl =
+                    m_enabledParallelForces->getElt(threadIndex-1)->getImpl();
+                if (impl.dependsOnlyOnPositions()) {
+                    impl.calcForce(*m_state, m_rigidBodyForceCacheLocal.upd(), m_particleForceCacheLocal.upd(), m_mobilityForceCacheLocal.upd());
+                } else { // ordinary velocity dependent force
+                    impl.calcForce(*m_state, m_rigidBodyForcesLocal.upd(), m_particleForcesLocal.upd(), m_mobilityForcesLocal.upd());
+                }
+            }
+            break;
+
+        case NonCached:
+            if (threadIndex == NonParallelForcesIndex) {
+                // Process all non-parallel forces.
+                for (Force* force : *m_enabledNonParallelForces) {
+                    const auto& impl = force->getImpl();
+                    if (!impl.dependsOnlyOnPositions()) {
+                        impl.calcForce(*m_state,
+                                *m_rigidBodyForces, *m_particleForces,
+                                *m_mobilityForces);
+                    }
+                }
+            } else {
+                // Process a single parallel force. Subtract 1 from index b/c
+                // we use 0 for the non-parallel forces.
+                const auto& impl =
+                    m_enabledParallelForces->getElt(threadIndex-1)->getImpl();
+                    if (!impl.dependsOnlyOnPositions()) {
+                        impl.calcForce(*m_state,
+                                m_rigidBodyForcesLocal.upd(), m_particleForcesLocal.upd(),
+                                m_mobilityForcesLocal.upd());
+                    }
+            }
+            break;
+        }
+    }
+    void finish() override {
+        // Add in this thread's contribution.
+        *m_rigidBodyForces += m_rigidBodyForcesLocal.get();
+        *m_particleForces += m_particleForcesLocal.get();
+        *m_mobilityForces += m_mobilityForcesLocal.get();
+
+        if (m_mode == CachedAndNonCached) {
+            *m_rigidBodyForceCache += m_rigidBodyForceCacheLocal.get();
+            *m_particleForceCache += m_particleForceCacheLocal.get();
+            *m_mobilityForceCache += m_mobilityForceCacheLocal.get();
+        }
+    }
+
+private:
+
+    ParallelExecutor& m_executor;
+
+    Mode m_mode;
+
+    ReferencePtr<const State> m_state;
+
+    // These are separated since we handle them differently.
+    const Array_<Force*>* m_enabledNonParallelForces;
+    const Array_<Force*>* m_enabledParallelForces;
```

Comment:
Should these be `std::unique_ptr<Array_<Force*>>`? Or do they even need to be heap allocated at all? How is ownership/destruction managed?